### PR TITLE
Let user specify heuristics: INDIRECT_OPTION and Max_Indirect_Target

### PR DESCRIPTION
**Pull request by bz-liu**
Let user specify heuristics: `INDIRECT_OPTION` and `Max_Indirect_Target`

### CODE REVIEW

File: src/PointerAnalysis/CMDOptions.cpp [hunk 1]
```diff
@@ -9,14 +9,21 @@ See the License for the specific language governing permissions and
...
 cl::opt<bool> USE_MEMLAYOUT_FILTERING(
     "Xmemlayout-filtering", cl::desc("Use memory layout to filter out incompatible types in field-sensitive PTA"));
 cl::opt<bool> CONFIG_VTABLE_MODE("Xenable-vtable", cl::desc("model vtable specially"), cl::init(false));
-cl::opt<bool> CONFIG_USE_FI_MODE("Xuse-fi-model", cl::desc("use field insensitive analyse"), cl::init(false));
\ No newline at end of file
+cl::opt<bool> CONFIG_USE_FI_MODE("Xuse-fi-model", cl::desc("use field insensitive analyse"), cl::init(false));
+
+// pta cmd options: set to default value
+IndirectResolveOption pta::INDIRECT_OPTION = IndirectResolveOption::WITH_LIMIT;
+unsigned int pta::Max_Indirect_Target = 999;
```
Comment (BradSwain) [inline src/PointerAnalysis/CMDOptions.cpp:29]: Just a small change, but can `999` be replaced `std::numeric_limits<unsigned int>::max()`. 
This way by default the max indirect target cap is effectively "off".

I think `#include <limits>` will need to be added as well.

File: src/PointerAnalysis/Models/LanguageModel/ConsGraphBuilder.h [hunk 2]
```diff
@@ -43,6 +43,8 @@ enum class IndirectResolveOption {
                // limit has been exceeded.
 };
 
+extern IndirectResolveOption INDIRECT_OPTION;  // default value = IndirectResolveOption::WITH_LIMIT
+
```
Comment (BradSwain) [inline src/PointerAnalysis/Models/LanguageModel/ConsGraphBuilder.h:47]: Is there anyway to pass this without relying on `extern`?

File: src/PointerAnalysis/Program/CtxFunction.h [hunk 3]
```diff
@@ -17,6 +17,8 @@ limitations under the License.
 
 namespace pta {
 
+extern unsigned int Max_Indirect_Target;  // the limit of the size of indirect targets -> default value 999
+
```
Comment (BradSwain) [inline src/PointerAnalysis/Program/CtxFunction.h:21]: Is there anyway to pass this without relying on `extern`?

File: src/main.cpp [hunk 4]
```diff
@@ -28,8 +28,8 @@ static llvm::cl::opt<std::string> DumpJSON("json", cl::desc("Dump JSON race repo
                                            cl::value_desc("destination file"));
 
 int main(int argc, char** argv) {
-  llvm::InitLLVM X(argc, argv);
-  llvm::cl::ParseCommandLineOptions(argc, argv);
+  // user input configs + llvm input args
+  pta::ParseCommandLineOptions(argc, argv);
```
Comment (BradSwain) [inline src/main.cpp:32]: I don't think we should move the command line option parsing into PTA.

If all the command line options for PTA are added as `llvm::cl` stuff they will be automatically parsed by `llvm::cl::ParseCommandLineOptions`

File: src/PointerAnalysis/Models/LanguageModel/ConsGraphBuilder.h [hunk 5]
```diff
@@ -224,7 +226,7 @@ class ConsGraphBuilder : public llvm::CtxInstVisitor<ctx, SubClass>, public PtrN
               }
 
               // if result == CRITICAL, do not apply limit
-              bool applyLimit = result == IndirectResolveOption::WITH_LIMIT;
+              bool applyLimit = (result == IndirectResolveOption::WITH_LIMIT ? true : false);
```
Comment (BradSwain) [inline src/PointerAnalysis/Models/LanguageModel/ConsGraphBuilder.h:229]: What does this change do?

**Review by BradSwain — CHANGES_REQUESTED** [review]

File: src/PointerAnalysis/CMDOptions.cpp (hunk 1 above)
Comment (bz-liu) [reply src/PointerAnalysis/CMDOptions.cpp:29]: Then what is `IndirectResolveOption::CRITICAL` represent? In the code it has a comment like: 
```c
enum class IndirectResolveOption {
  SKIP,        // do not add the resolved function to the callgraph
  WITH_LIMIT,  // only add the resolved function to the callgraph iff the
               // indirect call limit is not exceeded.
  CRITICAL,    // make sure the function are inserted to callgraph, even if the
               // limit has been exceeded.
};
```
Then both default `WITH_LIMIT` and `CRITICAL` will be effectively "off"

File: src/PointerAnalysis/Models/LanguageModel/ConsGraphBuilder.h (hunk 2 above)
Comment (bz-liu) [reply src/PointerAnalysis/Models/LanguageModel/ConsGraphBuilder.h:47]: I googled, did not find a solution. I'm not an c/c++ expert, maybe I googled in a wrong way? I see this `extern` has been used in the pta code, e.g., src/PointerAnalysis/Models/MemoryModel/FieldSensitive/MemBlock.h:25.

File: src/PointerAnalysis/Models/LanguageModel/ConsGraphBuilder.h (hunk 5 above)
Comment (bz-liu) [reply src/PointerAnalysis/Models/LanguageModel/ConsGraphBuilder.h:229]: previously, there is no option for `CRITICAL`, the commented-off line is the original code.

File: src/main.cpp (hunk 4 above)
Comment (bz-liu) [reply src/main.cpp:32]: ok, will update.

**Review by BradSwain — APPROVED** [review]

**Comment by BradSwain** [comment]
If possible we should try top refactor PTA to pass the configs in explicitly rather than relying on extern `llvm::cl` objects.
But not very high priority, so we can leave that for later.

**Comment by bz-liu** [comment]
I'll add a TODO in the code.

**Pull request merged**